### PR TITLE
feat(tencentcloud): Add security group for tencentcloud.

### PR DESCRIPTION
**Pull request by sean-3**
Add security group module for tencentcloud, 
Including creation, editing, deletion and details.

### CODE REVIEW

**Comment by sean-3** [comment]
@christopherthielen Please help to review this pr.

**Comment by christopherthielen** [comment]
I submitted a PR to your branch with some changed that demonstrate newer coding style:  https://github.com/TencentCloud/deck/pull/3

File: app/scripts/modules/tencentcloud/src/securityGroup/configure/CreateModal.tsx [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+  InfrastructureCaches,
+  NameUtils,
+} from '@spinnaker/core';
+import { Subject } from 'rxjs';
+import { get, uniq, intersection } from 'lodash';
+import { ISecurityGroup } from '@spinnaker/core';
+import { ISecurityGroupProps, ISecurityGroupDetail } from '../define';
+import { LoadBalancerLocation } from './components/LoadBalancerLocation';
+import { Ingress } from './components/Ingress';
+export interface ICreateSecurityGroupState {
```
Comment (christopherthielen) [inline app/scripts/modules/tencentcloud/src/securityGroup/configure/CreateModal.tsx:21]: It looks like a lot of this state isn't even used.  Please remove any state that is not necessary.  I think most of the state management is handled by the `Location` component.  (See the code in https://github.com/TencentCloud/deck/pull/3 )

File: app/scripts/modules/tencentcloud/src/securityGroup/configure/CreateModal.tsx [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    this.initializeSecurityGroups().then(this.initializeAccounts);
+  }
+
+  setSecurityGroupRefreshTime() {
+    this.setState({
+      refreshTime: InfrastructureCaches.get('securityGroups').getStats().ageMax,
+    });
+  }
+
+  private initializeSecurityGroups() {
```
Comment (christopherthielen) [inline app/scripts/modules/tencentcloud/src/securityGroup/configure/CreateModal.tsx:85]: I think equivalent code is now found in `FirewallLocationForm`

File: app/scripts/modules/tencentcloud/src/securityGroup/configure/CreateModal.tsx [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        allSecurityGroups: (securityGroups as unknown) as [],
+        // availableSecurityGroups: availableGroups.map((i: { name: string }) => i.name),
+      });
+      this.allSecurityGroupsUpdated.next();
+    });
+  }
+
+  private allSecurityGroupsUpdated = new Subject();
+
+  private initializeAccounts = () => {
```
Comment (christopherthielen) [inline app/scripts/modules/tencentcloud/src/securityGroup/configure/CreateModal.tsx:108]: I think equivalent code is now found in `FirewallLocationForm`

File: app/scripts/modules/tencentcloud/src/securityGroup/configure/CreateModal.tsx [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+      this.setState({ regions });
+      // clearSecurityGroups();
+      this.regionUpdated();
+      if (this.props.isNew) {
+        this.updateName();
+      }
+    });
+  }
+  updateName() {
+    const { securityGroup } = this.state;
```
Comment (christopherthielen) [inline app/scripts/modules/tencentcloud/src/securityGroup/configure/CreateModal.tsx:133]: I think equivalent code is now found in `FirewallLocationForm`

File: app/scripts/modules/tencentcloud/src/securityGroup/configure/CreateModal.tsx [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+      availableSecurityGroups,
+      existingSecurityGroupNames,
+      securityGroupsLoaded: true,
+    });
+  }
+  private regionUpdated() {
+    this.configureFilteredSecurityGroups();
+  }
+
+  addMoreItems() {
```
Comment (christopherthielen) [inline app/scripts/modules/tencentcloud/src/securityGroup/configure/CreateModal.tsx:170]: I am unsure what this is, but I don't think it's used.

File: app/scripts/modules/tencentcloud/src/securityGroup/configure/CreateModal.tsx [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+  private onTaskComplete = (): void => {
+    this.props.application.securityGroups.refresh();
+    this.refreshUnsubscribe = this.props.application.securityGroups.onNextRefresh(null, this.onApplicationRefresh);
+  };
+
+  public render() {
+    const { application, isNew, dismissModal } = this.props;
+    const { taskMonitor, securityGroup } = this.state;
+    return (
+      // @ts-ignore
```
Comment (christopherthielen) [inline app/scripts/modules/tencentcloud/src/securityGroup/configure/CreateModal.tsx:243]: Please remove `// @ts-ignore` and fix the types

File: app/scripts/modules/tencentcloud/src/securityGroup/define.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { Application, ISecurityGroup } from '@spinnaker/core';
```
Comment (christopherthielen) [inline app/scripts/modules/tencentcloud/src/securityGroup/define.ts:1]: rename this file to `interface.ts` to match the code style.

File: app/scripts/modules/tencentcloud/src/securityGroup/details/index.tsx [hunk 8]
```diff
@@ -0,0 +1,169 @@
+import * as React from 'react';
```
Comment (christopherthielen) [inline app/scripts/modules/tencentcloud/src/securityGroup/details/index.tsx:1]: `index.ts` files should contain only re-exports, not components.  Move this code to `SecurityGroupDetail.tsx`

File: app/scripts/modules/tencentcloud/src/vpc/VpcReader.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
+import { IPromise } from 'angular';
+
+import { IVpc, NetworkReader } from '@spinnaker/core';
+export interface ITencentcloudVpc extends IVpc {
+  isDefault?: boolean;
+}
+export class VpcReader {
```
Comment (christopherthielen) [inline app/scripts/modules/tencentcloud/src/vpc/VpcReader.ts:7]: `VPC` is an [Amazon Web Services term](https://aws.amazon.com/vpc/).  I see that it is also a [Tencent Cloud term](https://intl.cloud.tencent.com/product/vpc).  Are they the same thing?

Comment (sean-3) [reply app/scripts/modules/tencentcloud/src/vpc/VpcReader.ts:7]: yes.

**Comment by christopherthielen** [comment]
@sean-3 do you need more guidance on this PR?

**Comment by sean-3** [comment]
> @sean-3 do you need more guidance on this PR?
After code review, I haven't found any adjustments yet. Do you have any suggestions?

**Pull request closed without merging**